Bluetooth: GATT: Add support to notify by UUID

Vudentz opened the pull request:
This adds support to notify by UUID which can now be passed as parameter to bt_gatt_notify_cb, in addition to that bt_gatt_notify_cb can notify multiple values at once.

To test the notifications a new shell command is introduced, gatt notify, which test using bt_gatt_notify_cb with a callback and user_data but while doing this it uncover a problem with current design of TX callbacks which run in TX thread which has very limited stack size causing the shell to crash when printing something from within the callback, so this introduces a config option to defer the execution of TX callbacks to the system wq.

Vudentz committed:
Bluetooth: GATT: Add bt_gatt_foreach_attr_type

This adds bt_gatt_foreach_attr_type which can match attribute by UUID
and/or attribute user_data, in addition of that the user can also limit
the number of matches.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Make use of bt_gatt_foreach_attr_type on bt_gatt_notify

This make use of bt_gatt_foreach_attr_type to match the CCC UUID which
previously was not possible with bt_gatt_foreach_attr.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: shell: Make use of bt_gatt_foreach_attr_type

This make use of bt_gatt_foreach_attr_type in gatt show-db to match
attributes by UUID.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add support to notify by UUID

This reworks bt_gatt_notify_cb to allow passing an UUID, in addition to
that it can now accept multiple notification at once as there could be
multiple instance of the same UUID the user can set multiple set of
parameters.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Shell: Fix having gatt unregister twice

Only one command handle shall be needed.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Shell: Add gatt notify command

This adds gatt notify command which can be used to exercise the
bt_gatt_notify* APIs.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: conn: Defer bt_conn_tx callback to system wq

This makes the transmission complete callbacks to run on system wq
context so they are not executed in TX thread which usually has a much
smaller, and non-configurable, stack size.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1181,11 +1196,21 @@ void bt_conn_notify_tx(struct bt_conn *conn)
+#if defined(CONFIG_BT_CONN_DEFER_TX_CB)
+		/* Only submit if there is a callback set */
+		if (tx->data.cb) {
+			k_work_submit(&tx->work);
```
I fear this might cause deadlocks. IIRC the way this was designed is that we do the TX callbacks in the TX thread *before* trying to send new packets to the controller, so that when the thread comes to the point of sending more packets it's guaranteed to have the necessary TX contexts available (there's e.g. a code comment about this in the send_frag function)

jhedberg replied:
Actually, not a deadlock, but there's an `__ASSERT()` that'll be triggered in `add_pending_tx()` if/when this happens since this allocation is by design supposed to always succeed.

Vudentz replied:
Afaik the risk of the deadlock is already present in the current form since the callback can block as well, though I guess you are saying that when defering we need to make ensure there are enough context to continue since they are no longer freed immediately which I guess can be resolved by using k_fifo for free_tx so it can wait with k_fifo_get instead of asserting.

jhedberg replied:
> Afaik the risk of the deadlock is already present in the current form since the callback can block as well

@Vudentz no, since this is in the same thread, i.e. currently the allocation part doesn't happen until the callback returns.

> though I guess you are saying that when defering we need to make ensure there are enough context to continue since they are no longer freed immediately which I guess can be resolved by using k_fifo for free_tx so it can wait with k_fifo_get instead of asserting.

Basically this thread runs whenever there's a HCI Number of Completed Packets event. Each "packet" maps to a tx context, so the idea is that this context is freed up for reuse by the HCI event. Using a k_fifo for the tx contexts and then blocking in the TX thread would indeed work, but then we need a k_fifo (which takes more memory) and we also need to extend the tx context with 4 bytes for the k_fifo (which also takes more memory), so either way we're looking at increasing the memory footprint. Making the TX thread size user configurable is at least complexity-wise the simpler option since that doesn't require adding any new code.

Vudentz replied:
> > though I guess you are saying that when defering we need to make ensure there are enough context to continue since they are no longer freed immediately which I guess can be resolved by using k_fifo for free_tx so it can wait with k_fifo_get instead of asserting.
> 
> Basically this thread runs whenever there's a HCI Number of Completed Packets event. Each "packet" maps to a tx context, so the idea is that this context is freed up for reuse by the HCI event. Using a k_fifo for the tx contexts and then blocking in the TX thread would indeed work, but then we need a k_fifo (which takes more memory) and we also need to extend the tx context with 4 bytes for the k_fifo (which also takes more memory), so either way we're looking at increasing the memory footprint. Making the TX thread size user configurable is at least complexity-wise the simpler option since that doesn't require adding any new code.

So I did the change to k_fifo which essencially uses the same node, since it is either on tx_free or tx_pending, I think this is the safest bet since we don't incor in using more memory and have the safety of using k_fifo_get which does properly wait for the callbacks to complete.

Vudentz replied:
Btw, it seems I one would have substantially increase the stack size in order to have something like a shell_print on the callback, which is why I continue with system wq approach.

Anyway the use of k_fifo_get is propably safer even when not defering so even if we were to drop the use of system wq I think it best to keep the use of k_fifo for tx_free.

jhedberg replied:
Yes, makes sense. If forgot the conn_tx contexts already had the necessary space for being inserted into a k_fifo.

jarz-nordic commented on subsys/bluetooth/shell/gatt.c:
```diff
@@ -525,21 +525,9 @@ static struct db_stats {
-		}
-		return BT_GATT_ITER_CONTINUE;
-	}
+	struct shell *shell = user_data;
```
I think it should be `const struct shell *shell = user_data;`

Vudentz replied:
Fixed.

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -70,6 +70,7 @@ config BT_HCI_TX_PRIO
 	int
 	default 7
 
+
```
This seems unnecessary

Vudentz replied:
Indeed, will fix it

jhedberg replied:
Seems it's still there?

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -183,6 +184,13 @@ config BT_CONN_TX_MAX
 	  been acknowledged by the controller.
 
+config BT_CONN_DEFER_TX_CB
+	bool "Defer transmission complete callback to system workqueue"
```
Making this configurable doesn't feel like the right solution (or at least sufficient). It has been an oversight *not* to have the TX thread size app-configurable, yet still have a callback from the thread going to the app. So either this option should eliminate the callback support all together, or then the TX thread size should be made app-configurable. Simplest solution would be to just always do the deferring and remove this new option.

Vudentz replied:
Sure, I was debating about this, just thought Id use a lot more elements on bt_conn_tx but it end up just be 8 bytes (56 bytes in total with current default of BT_CONN_TX_MAX).

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -746,35 +769,51 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
+	const void *data;
+	/** Notification Value length */
+	u16_t len;
+	/** Notification Value callback */
```
This should clearly document the context (thread) that the callback will be called from. We should probably do a review of all our APIs to make sure it's clearly documented what context various callbacks get called from.

Vudentz replied:
Right will add the context of the callbacks.

Vudentz replied:
Fixed

jhedberg replied:
Wouldn't it be better to have a mention of it here as well, since that's likely the first place a developer will look. In fact, this probably deserves a bit more explanation about when exactly the callback happens (i.e. what it means).

Vudentz committed:
Bluetooth: GATT: Disable flow control if running with POSIX_ARCH

Disable flow control with POSIX_ARCH since these boards tend to run
faster than normal defering the TX callback to system wq may not be
able to run before another packet is received.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Document the context in which callbacks are run

This documents the context in which callbacks are run, that is now
the System Workqueue.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1205,45 +1220,43 @@ static void notify_tx(void)
-	node = sys_slist_get_not_empty(&free_tx);
-	tx = CONTAINER_OF(node, struct bt_conn_tx, node);
+	tx = k_fifo_get(&free_tx, K_FOREVER);
+	tx->conn = conn;
```
This should actually be `tx->conn = bt_conn_ref(conn)` and then the corresponding `unref` in the `tx_free()` function

Vudentz replied:
Would this prevent the conn to be free if that are still refs to the conn, I though that could prevent conn_cleanup to be run or that just depend on the DISCONNECTED state?

jhedberg replied:
You shouldn't confuse the connection state with connection reference counting. Those are completely independent things. The reference counting is simply counting how many pointers variables you have pointing at the same conn object, regardless of what the state is.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -287,6 +287,26 @@ static void conn_le_update_timeout(struct k_work *work)
 
+static void tx_free(struct bt_conn_tx *tx)
+{
+	tx->conn = NULL;
```
This should be:
```
if (tx->conn) {
        bt_conn_unref(tx->conn);
        tx->conn = NULL;
}
```
see my other comment about where the `ref()` is needed

Vudentz replied:
Fixed.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1475,20 +1499,45 @@ static u8_t notify_cb(const struct bt_gatt_attr *attr, void *user_data)
-		      bt_gatt_complete_func_t func, void *user_data)
+static u8_t match_uuid(const struct bt_gatt_attr *attr, void *user_data)
+{
+	struct bt_gatt_attr **found = user_data;
```
Shouldn't this variable type have a `const` somewhere so you don't need the typecast below?

Vudentz replied:
Right, will fix it.

Vudentz replied:
Fixed.

aescolar commented:
CC @wdhpawa

Vudentz commented:
@jhedberg I guess we can apply this one since there hasn't been any more comments.

jhedberg commented:
> @jhedberg I guess we can apply this one since there hasn't been any more comments.

@Vudentz done
